include/LightGBM/utils/common.h

```diff
@@ -1082,12 +1082,18 @@ struct __StringToTHelper<T, true> {
     // Fast (common) path: For numeric inputs in RFC 7159 format:
     const bool fast_parse_succeeded = fast_double_parser::parse_number(str.c_str(), &tmp);
 
-    // Rare path: Not in RFC 7159 format. Possible "inf", "nan", etc. Fallback to standard library:
+    // Rare path: Not in RFC 7159 format. Possible "inf", "nan", etc.
     if (!fast_parse_succeeded) {
-      std::stringstream ss;
-      Common::C_stringstream(ss);
-      ss << str;
-      ss >> tmp;
```

Comment:
Keep this code on the `else` branch instead of raising a fatal error.

Reply:
But that code silently fails, which is completely unacceptable.

In reality we are only parsing strings generated by LightGBM itself when the model was written out to file, so we should ensure there are robust round-trip tests which include models that contain inf and nan values. There are no other possible non-finite values defined for IEEE 754 floating point numbers, and if at some point in the future the standard changed, this would be quickly picked up by the round-trip tests and easily addressed.

I am shocked that such tests don't already exist.

Reply:
I meant doing number parsing after the nan and inf checks. Your logic makes sense regarding IEEE-754, although I wouldn't recommend dropping that parsing at the end without adding said tests first, not 100.0% sure fast_double_parser parses all numbers.